drivers/platform/maxim/max32650/maxim_usb_uart.c

```diff
@@ -0,0 +1,509 @@
+ *  - Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ *  - Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ *  - Neither the name of Analog Devices, Inc. nor the names of its
+ *    contributors may be used to endorse or promote products derived
+ *    from this software without specific prior written permission.
+ *  - The use of this software may or may not infringe the patent rights
+ *    of one or more patent holders.  This license does not release you
+ *    from the requirement that you obtain separate licenses from these
+ *    patent holders to use this software.
+ *  - Use of the software either in source or binary form, must be run
+ *    on or directly connected to an Analog Devices Inc. component.
+ *
+ * THIS SOFTWARE IS PROVIDED BY ANALOG DEVICES "AS IS" AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, NON-INFRINGEMENT,
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL ANALOG DEVICES BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, INTELLECTUAL PROPERTY RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+#include <errno.h>
+#include <stdlib.h>
+#include <stddef.h>
+#include "maxim_irq.h"
+#include "maxim_usb_uart.h"
+#include "no_os_irq.h"
+#include "no_os_util.h"
+#include "no_os_delay.h"
+#include "no_os_alloc.h"
+#include "no_os_lf256fifo.h"
+#include "mxc_errors.h"
+#if (TARGET_NUM != 32650)
+#include "mcr_regs.h"
+#endif
+#include "mxc_sys.h"
+#include "mxc_delay.h"
+#include "usb.h"
+#include "usb_event.h"
+#include "enumerate.h"
+#include "cdc_acm.h"
+#include "maxim_usb_uart_descriptors.h"
+
+#define EVENT_ENUM_COMP MAXUSB_NUM_EVENTS
+#define EVENT_REMOTE_WAKE (EVENT_ENUM_COMP + 1)
+
+static int setconfig_callback(MXC_USB_SetupPkt *sud, void *cbdata);
+static int setfeatureCallback(MXC_USB_SetupPkt *sud, void *cbdata);
+static int clrfeatureCallback(MXC_USB_SetupPkt *sud, void *cbdata);
+static int event_callback(maxusb_event_t evt, void *data);
+static void usbAppSleep(void);
+static void usbAppWakeup(void);
+static int usb_read_callback(void);
+int usbStartupCallback();
+int usbShutdownCallback();
+
+static void usb_callback(void *context)
+{
+	MXC_USB_EventHandler();
```

Comment:
A minor remark, but I think this function should be called in `USB_IRQHandler()`. I assume that it's responsible for clearing the NVIC interrupt bit for the USB interrupt. If you do this change, `usb_callback()` may be removed and then it wouldn't be required to register an interrupt.

Reply:
I could but I was thinking that you might wanna use CDC ACM class as well as let's say file storage class on the same USB in which case (I think, I have not documented myself) there would be at least 2 things to call from USB_IRQHandler so you could hook up 2 respective callbacks with our no_os_irq infrastructure.